huntsman/observatory.py

```diff
@@ -232,202 +233,67 @@ def autofocus_cameras(self, *args, **kwargs):
-                    break
-
-                camera = self.cameras[cam_name]
-
-                filename = "{}/flats/{}/{}/{}.{}".format(
-                    image_dir,
-                    camera.uid,
-                    flat_obs.seq_time,
-                    'dark_{:02d}'.format(flat_obs.current_exp_num),
-                    camera.file_extension)
-
-                # Take picture and wait for result
-                camera_event = camera.take_observation(
-                    flat_obs,
-                    fits_headers,
-                    filename=filename,
-                    exptime=exptime,
-                    dark=True
-                )
-
-                camera_events[cam_name] = {
-                    'event': camera_event,
-                    'filename': filename,
-                }
-
-            # Will block here until done exposing on all cameras
-            while not all([info['event'].is_set() for info in camera_events.values()]):
-                self.logger.debug('Waiting for dark-field image')
-                time.sleep(1)
-
-    def take_morning_flats(*args, **kwargs):
-        '''
-
-        '''
-        raise NotImplementedError('Morning flats not implemented yet.')
+        if camera_names is None:
+            cameras_all = self.cameras
+        else:
+            cameras_all = {c: self.cameras[c] for c in camera_names}
+
+        # Obtain the filter order
+        filter_order = self.config['filterwheel']['flat_field_order']
+        if self.past_midnight:  # If it's the morning, order is reversed
+            filter_order = filter_order[::-1]
+
+        exptimes_dark = {c: [0 * u.second] for c in cameras_all.keys()}  # Add a bias exposure
+        for filter_name in filter_order:
+
+            # Get a dict of cameras that have this filter
+            cameras = {}
+            for cam_name, cam in cameras_all.items():
+                if cam.filterwheel is None:
+                    if cam.filter_type == filter_name:
+                        cameras[cam_name] = cam
+                elif filter_name in cam.filterwheel.filter_names:
+                    cameras[cam_name] = cam
+
+            # Go to next filter if there are no cameras with this one
+            if not cameras:
+                self.logger.debug(f'No cameras found with {filter_name} filter.')
+                continue
+
+            # Create the Observation object
+            obs = self._create_flat_field_observation(alt=alt, az=az, filter_name=filter_name)
+
+            # Take the flats for each camera in this filter
+            self.logger.info(f'Taking flat fields in {filter_name} filter.')
+            exptimes = self._take_autoexposure_flats(cameras, obs, **kwargs)
+
+            # Log the new exposure times we need to take darks with
+            for cam_name in cameras.keys():
+                exptimes_dark[cam_name].extend(
+                    [e for e in exptimes[cam_name] if e not in exptimes_dark[cam_name]])
```

Comment:
If you use `set`s instead of `list`s (see my other suggestion) you don't have to do this filtering yourself.

```suggestion
                exptimes_dark[cam_name].update(exptimes[cam_name])
```